cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java

```diff
@@ -334,37 +472,105 @@ private boolean updateV1Service(CoreV1Api api, V1Service currentService,
     service.getSpec().setPorts(Collections.singletonList(port));
     service.getSpec().setSelector(podLabels);
 
-    try {
-      api.replaceNamespacedService(meta.getName(), namespace, service, null, null, null, null);
-      LOG.info("Service updated in kubernetes with name {} and port {}",
-          currentService.getMetadata().getName(), port.getPort());
-    } catch (ApiException e) {
-      if (e.getCode() == HttpURLConnection.HTTP_CONFLICT
-          || e.getCode() == HttpURLConnection.HTTP_NOT_FOUND) {
-        return false;
+    // Update service type.
+    if (SERVICE_TYPE_LOAD_BALANCER.equals(getServiceType(discoverable))) {
+      service.getSpec().setType(SERVICE_TYPE_LOAD_BALANCER);
+      Map<String, String> annotations = new HashMap<>(meta.getAnnotations());
+      annotations.putAll(loadBalancerServiceAnnotations);
+      meta.setAnnotations(annotations);
+    } else {
+      service.getSpec().setType(SERVICE_TYPE_CLUSTER_IP);
+      // Remove load balancer annotations.
+      if (meta.getAnnotations() != null) {
+        Map<String, String> annotations = new HashMap<>(meta.getAnnotations());
+        annotations.keySet().removeAll(loadBalancerServiceAnnotations.keySet());
+        meta.setAnnotations(annotations);
       }
-      throw e;
     }
-    return true;
+
+    return service;
   }
 
+
   /**
-   * Closes a {@link AutoCloseable} and swallow any exception.
+   * Creates a {@link Set} of {@link Discoverable} for the given service.
    *
-   * @param closeable if not null, the {@link AutoCloseable#close()} of the given closeable will
-   *     be called
+   * @param name    name of the service
+   * @param service the K8s service object for creating the Discoverable
+   * @return a {@link Set} of {@link Discoverable}.
    */
-  private void closeQuietly(@Nullable AutoCloseable closeable) {
-    if (closeable == null) {
-      return;
+  @VisibleForTesting
+  Set<Discoverable> toDiscoverables(String name, V1Service service,
+      String namespace) {
+    V1ObjectMeta meta = service.getMetadata();
+    String hostname = meta.getName();
+    List<V1ServicePort> servicePorts = service.getSpec().getPorts();
+
+    // Decode the payload from annotation. If absent, default to empty payload
+    byte[] payload = Optional.ofNullable(meta.getAnnotations())
+        .map(m -> m.get(PAYLOAD_NAME))
+        .map(Base64.getDecoder()::decode)
+        .orElse(EMPTY_PAYLOAD);
+
+    // We don't expect there is more than one service port, hence only pick the first one
+    Stream<Discoverable> discoverableStream;
+    if (SERVICE_TYPE_LOAD_BALANCER.equals(service.getSpec().getType())) {
+      Optional<String> ipAddr = getLoadBalancerIp(service);
+      if (!ipAddr.isPresent()) {
+        LOG.warn(
+            "Didn't find any IP addresses for Load Balancer service {}. No discoverable created.",
+            name);
+        return Collections.emptySet();
+      }
+      discoverableStream = servicePorts.stream()
```

Comment:
nit:
can use ternary operator to derive discoverable name and just make `createDiscoverable` call during return statement.

Reply:
Instead of making a stream in the if/else block, I've created only the hostname. The creation of stream by calling `createDiscoverable()` is common now so its outside the conditional block.